File: internal/installer/brewcask/brewcask_test.go
package brewcask_test

import (
	"context"
	"testing"

	"gearup/internal/config"
	gearexec "gearup/internal/exec"
	"gearup/internal/installer/brewcask"
)

func TestBrewCask_CheckInstalled(t *testing.T) {
	f := gearexec.NewFakeRunner()
	f.On("brew list --cask iterm2").Return(gearexec.Result{ExitCode: 0}, nil)

	inst := &brewcask.Installer{Runner: f}
	ok, err := inst.Check(context.Background(), config.Step{Name: "iTerm2", Type: "brew-cask", Cask: "iterm2"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !ok {
		t.Error("installed = false, want true")
	}
}

func TestBrewCask_CheckNotInstalled(t *testing.T) {
	f := gearexec.NewFakeRunner()
	f.On("brew list --cask iterm2").Return(gearexec.Result{ExitCode: 1, Stderr: "Error: Cask iterm2 is not installed"}, nil)

	inst := &brewcask.Installer{Runner: f}
	ok, err := inst.Check(context.Background(), config.Step{Name: "iTerm2", Type: "brew-cask", Cask: "iterm2"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if ok {
		t.Error("installed = true, want false")
	}
}

func TestBrewCask_CheckMissingCask(t *testing.T) {
	f := gearexec.NewFakeRunner()
	inst := &brewcask.Installer{Runner: f}
	_, err := inst.Check(context.Background(), config.Step{Name: "bad", Type: "brew-cask"})
	if err == nil {
		t.Error("want error for missing cask, got nil")
	}
}

func TestBrewCask_CheckUsesOverride(t *testing.T) {
	f := gearexec.NewFakeRunner()
	f.On("test -d /Applications/iTerm.app").Return(gearexec.Result{ExitCode: 0}, nil)

	inst := &brewcask.Installer{Runner: f}
	ok, err := inst.Check(context.Background(), config.Step{
		Name: "iTerm2", Type: "brew-cask", Cask: "iterm2",
		Check: "test -d /Applications/iTerm.app",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !ok {
		t.Error("installed = false, want true")
	}
	if got := f.Calls(); len(got) != 1 || got[0].Cmd != "test -d /Applications/iTerm.app" {
		t.Errorf("Calls = %+v, want override command", got)
	}
}

func TestBrewCask_InstallSuccess(t *testing.T) {
	f := gearexec.NewFakeRunner()
	f.On("brew install --cask iterm2").Return(gearexec.Result{ExitCode: 0}, nil)

	inst := &brewcask.Installer{Runner: f}
	err := inst.Install(context.Background(), config.Step{Name: "iTerm2", Type: "brew-cask", Cask: "iterm2"})
	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestBrewCask_InstallFailure(t *testing.T) {
	f := gearexec.NewFakeRunner()
	f.On("brew install --cask badcask").Return(gearexec.Result{ExitCode: 1, Stderr: "Error: Cask badcask not found"}, nil)

	inst := &brewcask.Installer{Runner: f}
	err := inst.Install(context.Background(), config.Step{Name: "bad", Type: "brew-cask", Cask: "badcask"})
	if err == nil {
		t.Error("want error for failed install, got nil")
	}
}

File: src/steps/brew-cask.test.ts
import { describe, it, expect } from "bun:test"
import { checkBrewCask, installBrewCask } from "./brew-cask"
import { FakeExec } from "../exec/fake"
import { makeContext } from "../context"
import type { BrewCaskStep } from "../schema"

describe("checkBrewCask", () => {
  it("runs the default `brew list --cask` check", async () => {
    const exec = new FakeExec()
    exec.queueResponse({ exitCode: 0 })
    const ctx = makeContext({ exec })

    const step: BrewCaskStep = { type: "brew-cask", name: "iTerm2", cask: "iterm2" }
    const result = await checkBrewCask(step, ctx)

    expect(result.installed).toBe(true)
    expect(exec.calls[0]?.argv).toEqual(["brew", "list", "--cask", "iterm2"])
    expect(exec.calls[0]?.shell).toBeFalsy()
  })

  it("returns installed=false on non-zero exit", async () => {
    const exec = new FakeExec()
    exec.queueResponse({ exitCode: 1 })
    const ctx = makeContext({ exec })

    const step: BrewCaskStep = { type: "brew-cask", name: "iTerm2", cask: "iterm2" }
    const result = await checkBrewCask(step, ctx)

    expect(result.installed).toBe(false)
  })

  it("runs step.check in shell when provided", async () => {
    const exec = new FakeExec()
    exec.queueResponse({ exitCode: 0 })
    const ctx = makeContext({ exec })

    const step: BrewCaskStep = {
      type: "brew-cask",
      name: "iTerm2",
      cask: "iterm2",
      check: 'test -d "/Applications/iTerm.app"',
    }
    const result = await checkBrewCask(step, ctx)

    expect(exec.calls[0]?.argv).toEqual(['test -d "/Applications/iTerm.app"'])
    expect(exec.calls[0]?.shell).toBe(true)
    expect(result.installed).toBe(true)
  })
})

describe("installBrewCask", () => {
  it("runs `brew install --cask <cask>` and returns ok on exit 0", async () => {
    const exec = new FakeExec()
    exec.queueResponse({ exitCode: 0 })
    const ctx = makeContext({ exec })

    const step: BrewCaskStep = { type: "brew-cask", name: "iTerm2", cask: "iterm2" }
    const result = await installBrewCask(step, ctx)

    expect(result.ok).toBe(true)
    expect(exec.calls[0]?.argv).toEqual(["brew", "install", "--cask", "iterm2"])
    expect(exec.calls[0]?.shell).toBeFalsy()
  })

  it("returns ok=false with error when exit code is non-zero", async () => {
    const exec = new FakeExec()
    exec.queueResponse({ exitCode: 1, stderr: "Cask not found" })
    const ctx = makeContext({ exec })

    const step: BrewCaskStep = { type: "brew-cask", name: "missing", cask: "missing" }
    const result = await installBrewCask(step, ctx)

    expect(result.ok).toBe(false)
    if (!result.ok) {
      expect(result.error).toContain("brew install --cask missing")
      expect(result.error).toContain("exit 1")
      expect(result.error).toContain("Cask not found")
    }
  })
})
